File: api/check.py
from http.server import BaseHTTPRequestHandler
import os
import requests
import sqlite3
import psycopg2
from urllib.parse import urlparse

# 環境變數
LINE_CHANNEL_ACCESS_TOKEN = os.environ.get('LINE_CHANNEL_ACCESS_TOKEN')
DATABASE_URL = os.environ.get('POSTGRES_URL') # Vercel Postgres usually provides this

def get_db_connection():
    if DATABASE_URL:
        # Use Postgres
        conn = psycopg2.connect(DATABASE_URL)
    else:
        # Fallback to local SQLite (Not persistent on Vercel!)
        conn = sqlite3.connect('waitlist.db')
    return conn

def check_inline_availability(company_id, branch_id, date, party_size):
    url = f"https://inline.app/api/companies/{company_id}/branches/{branch_id}/capacities?date={date}&partySize={party_size}"
    try:
        resp = requests.get(url, headers={"User-Agent": "Mozilla/5.0"}, timeout=10)
        if resp.status_code == 200:
            data = resp.json()
            # 這裡簡化邏輯，真實情況需解析 JSON 結構
            # 假設 API 回傳 available: true 或類似結構
            # 這裡回傳模擬數據，若要真實運作需配合 Proxy
            return [] # ["18:00"] 
    except:
        pass
    return []

def push_line_notification(user_id, message):
    if not LINE_CHANNEL_ACCESS_TOKEN:
        print("Missing LINE Token")
        return
        
    url = "https://api.line.me/v2/bot/message/push"
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {LINE_CHANNEL_ACCESS_TOKEN}"
    }
    data = {
        "to": user_id,
        "messages": [{"type": "text", "text": message}]
    }
    requests.post(url, headers=headers, json=data)

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        print("🚀 Cron Job Triggered")
        try:
            conn = get_db_connection()
            cursor = conn.cursor()
            
            # 1. 撈出 PENDING 任務
            # 注意: table schema 需與資料庫一致
            query = "SELECT t.id, t.user_id, t.target_date, t.party_size, r.company_id, r.branch_id, r.booking_url FROM tasks t JOIN restaurants r ON t.restaurant_id = r.id WHERE t.status='PENDING'"
            cursor.execute(query)
            tasks = cursor.fetchall()
            
            results = []
            
            for task in tasks:
                task_id, uid, date, size, cid, bid, link = task
                
                # 2. 檢查空位
                slots = check_inline_availability(cid, bid, date, size)
                
                if slots:
                    msg = f"🔥 發現空位！\n日期：{date}\n時段：{slots}\n快搶：{link}"
                    push_line_notification(uid, msg)
                    
                    # 更新狀態
                    update_query = "UPDATE tasks SET status='FOUND' WHERE id=%s" if DATABASE_URL else "UPDATE tasks SET status='FOUND' WHERE id=?"
                    cursor.execute(update_query, (task_id,))
                    conn.commit()
                    results.append(f"Notified {uid}")
            
            cursor.close()
            conn.close()
            
            self.send_response(200)
            self.end_headers()
            self.wfile.write(f"Checked {len(tasks)} tasks. Results: {results}".encode())
            
        except Exception as e:
            print(f"Error: {e}")
            self.send_response(500)
            self.end_headers()
            self.wfile.write(str(e).encode())

File: api/check.ts
import { sql } from '@vercel/postgres';
import { VercelRequest, VercelResponse } from '@vercel/node';
import * as line from '@line/bot-sdk';

const config = {
    channelAccessToken: process.env.LINE_CHANNEL_ACCESS_TOKEN || '',
    channelSecret: process.env.LINE_CHANNEL_SECRET || '',
};

const client = new line.Client(config);

async function checkAvailability(url: string, targetDate: string, partySize: number) {
    try {
        const response = await fetch(url);
        const html = await response.text();

        // This logic mimics the Python one: finding window.__INITIAL_STATE__
        const prefix = 'window.__INITIAL_STATE__ = ';
        const jsonStart = html.indexOf(prefix);
        if (jsonStart !== -1) {
            // Simple string extraction to check for "available"
            // In a real robust implementation we might parse JSON, but here we can just check existence
            // Logic: If we see available times for the target date, return true
            // For simplify: just return true for test if date is found in response
            return html.includes(targetDate);
        }
    } catch (e) {
        console.error("Check Error:", e);
    }
    return false; // Assume false if error
}

export default async function handler(req: VercelRequest, res: VercelResponse) {
    console.log("🚀 Cron Job Triggered (Node.js)");

    try {
        // 1. Fetch PENDING tasks
        const { rows: tasks } = await sql`
            SELECT t.id, t.user_id, t.target_date, t.party_size, r.company_id, r.branch_id, r.booking_url 
            FROM tasks t 
            JOIN restaurants r ON t.restaurant_id = r.id 
            WHERE t.status='PENDING'
        `;

        const results = [];

        for (const task of tasks) {
            // 2. Check Availability
            // In a real crawler, we would call the inline API APIs or parse the page
            // Here we use a placeholder check logic or the previous mock logic

            // NOTE: Since the previous Python code was using requests to fetch page content, 
            // we will simulate the check here.

            const isAvailable = await checkAvailability(task.booking_url, task.target_date, task.party_size);

            if (isAvailable) {
                // 3. Send Notification
                try {
                    await client.pushMessage(task.user_id, {
                        type: 'text',
                        text: `🔥 發現空位！\n餐廳：${task.url}\n日期：${task.target_date}\n人數：${task.party_size}人\n請盡快訂位！`
                    });

                    // 4. Update Status
                    await sql`UPDATE tasks SET status='NOTIFIED' WHERE id=${task.id}`;
                    results.push(`Notified user ${task.user_id}`);
                } catch (lineError) {
                    console.error("LINE Error:", lineError);
                }
            } else {
                results.push(`Task ${task.id}: No availability`);
            }
        }

        res.status(200).json({ success: true, results });

    } catch (error) {
        console.error(error);
        res.status(500).json({ error: String(error) });
    }
}
